File: code/graph-algorithms/diameter_trees/cpp/Diameter.cpp
//Learn O(n)


// The following is O(n^2)
#include<bits/stdc++.h>
#define ll long long
using namespace std;

typedef struct tree node;

struct tree
{
	ll data;
	node * right;
	node * left;
};

node * insert(ll a)
{
	tree * temp = new tree;
	temp->data=a;
	temp->left=temp->right=NULL;
	return temp;
}

//Height including the root
ll height(node * root)
{
	if(root==NULL)
	{
		return 0;
	}
	else return 1 + max(height(root->left),height(root->right));
}

// Diameter ( checking for all nodes)
ll diameter(node * root)
{
	if(root==NULL)
	{
		return 0;
	}
	else return max(1 + height(root->right) + height(root->left),diameter(root->right)+diameter(root->left));
}
	

int main()
{
	//The tree;
	node * root=insert(1);
	//cout<<height(root);
	root->left=insert(3);
	root->left->left=insert(4);
	root->left->right=insert(5);
	root->left->left->right=insert(6);
	root->left->right->right=insert(7);
	
	
	/*tree * root=new tree;
	root=NULL;
	*/
	cout<<diameter(root);	
        return 0;
}

File: code/data_structures/src/tree/binary_tree/binary_tree/diameter/diameter.py
# Time:  O(n)
# Space: O(h)

# Given a binary tree, you need to compute the length of the diameter of the tree.
# The diameter of a binary tree is the length of the longest path between
# any two nodes in a tree. This path may or may not pass through the root.
#
# Example:
# Given a binary tree
#           1
#          / \
#         2   3
#       / \
#       4   5
# Return 3, which is the length of the path [4,2,1,3] or [5,2,1,3].
#
# Note: The length of path between two nodes is represented by the number of edges between them.

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


# The function diameter_height returns the diameter and the height of the tree.
# And the function find_tree_diameter uses it to just compute the diameter (by discarding the height).
class Solution:
    def diameter_height(node):
        if node is None:
            return 0, 0
        ld, lh = diameter_height(node.left)
        rd, rh = diameter_height(node.right)
        return max(lh + rh + 1, ld, rd), 1 + max(lh, rh)

    def find_tree_diameter(node):
        d, _ = diameter_height(node)
        return d


diameter = Solution()
